Packages/com.unity.inputsystem/InputSystem/Editor/ProjectWideActions/ProjectWideActionsAsset.cs

```diff
@@ -97,10 +105,51 @@ private static InputActionAsset CreateNewActionAsset()
                     AssetDatabase.AddObjectToAsset(actionReference, asset);
                 }
             }
+        }
 
-            AssetDatabase.SaveAssets();
+        /// <summary>
+        /// Updates the input action references in the asset by updating names, removing dangling references
+        /// and adding new ones.
+        /// </summary>
+        /// <param name="asset"></param>
+        internal static void UpdateInputActionReferences()
+        {
+            var asset = GetOrCreate();
+            var existingReferences = InputActionImporter.LoadInputActionReferencesFromAsset(asset).ToList();
 
-            return asset;
+            // Check if referenced input action exists in the asset and remove the reference if it doesn't.
+            foreach (var actionReference in existingReferences)
+            {
+                var action = asset.FindAction(actionReference.action.id);
+                if (action == null)
+                {
+                    actionReference.Set(null);
+                    AssetDatabase.RemoveObjectFromAsset(actionReference);
+                }
+            }
+
+            // Check if all actions have a reference
+            foreach (var action in asset)
+            {
+                var actionReference = existingReferences.FirstOrDefault(r => r.m_ActionId == action.id.ToString());
```

Comment:
Feels like we could do this faster, e.g. by generating a hash in the above foreach look and then checking if its in the hash here - rather than splining the eixtingReferences list on every action in the asset.

Reply:
Thanks. We probably can. I'll leave it for the end in case the direction in this PR is accepted.
Are you suggesting to create some kind of dictionary to associate the action ID hash of the reference with the reference itself ? So that we can then use the `action.id` here as a key ?